<?xml version="1.0"?>
<!--
  Copyright (C) 2019  Rhys Mainwaring

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <https://www.gnu.org/licenses/>.
-->
<sdf version="1.6">
  <model name="wing_sail_with_elevator">
    <pose>0 0 0.05 0 0 0</pose>
    <static>false</static>

    <!-- Sail plugin -->
    <!-- <plugin name="wing_sail_liftdrag" filename="libSailPlugin.so">
        <a0>0.0</a0>
        <cla>6.2832</cla>
        <alpha_stall>0.1592</alpha_stall>
        <cla_stall>-0.7083</cla_stall>
        <cda>0.63662</cda>
        <area>0.4858</area>
        <fluid_density>1.2</fluid_density>
        <forward>1 0 0</forward>
        <upward>0 1 0</upward>
        <cp>-0.05 0.0 0.65</cp>
        <link_name>wing_sail_link</link_name>
        <radial_symmetry>true</radial_symmetry>
        <topic>lift_drag</topic>
    </plugin> -->

    <!-- Gazebo / ROS plugins -->

    <!-- joint states -->
    <!-- <plugin name="gazebo_ros_joint_state_publisher" filename="libgazebo_ros_joint_state_publisher.so">
        <robotNamespace>/</robotNamespace>
        <jointName>
            wing_sail_joint
        </jointName>
        <updateRate>50</updateRate>
    </plugin> -->

    <!-- ground truth -->
    <!-- <plugin name="gazebo_ros_f3d" filename="libgazebo_ros_f3d.so">
        <robotNamespace>/</robotNamespace>
        <bodyName>wing_sail_link</bodyName>
        <topicName>/gazebo/f3d/wing_sail_link</topicName>
        <frameName>world</frameName>
        <updateRate>50</updateRate>
    </plugin> -->

    <!-- force torque sensor -->
    <!-- <plugin name="gazebo_ros_ft_sensor" filename="libgazebo_ros_ft_sensor.so">
        <robotNamespace>/</robotNamespace>
        <jointName>wing_sail_joint</jointName>
        <topicName>/gazebo/ft_sensor/wing_sail_joint</topicName>
        <gaussianNoise>0</gaussianNoise>
        <updateRate>50</updateRate>
    </plugin> -->

    <!-- Base -->
    <link name="base_link">
      <visual name="visual">
        <geometry>
          <box>
            <size>0.5 0.5 0.1</size>
          </box>
        </geometry>
        <material>
          <ambient>0.5 0.5 0.5 1.0</ambient>
          <diffuse>0.5 0.5 0.5 1.0</diffuse>
          <specular>0.1 0.1 0.1 1.0</specular>
        </material>
      </visual>
      <collision name="collision">
        <geometry>
          <box>
            <size>0.5 0.5 0.1</size>
          </box>
        </geometry>
      </collision>

      <visual name="imu_visual">
        <geometry>
          <box>
            <size>0.04 0.04 0.01</size>
          </box>
        </geometry>
        <material>
          <ambient>0.0 1.0 0.0 1.0</ambient>
          <diffuse>0.0 1.0 0.0 1.0</diffuse>
          <specular>0.1 0.1 0.1 1.0</specular>
        </material>
      </visual>

      <sensor name="imu_sensor" type="imu">
        <pose>0 0 0 3.141593 0 0</pose>
        <always_on>1</always_on>
        <update_rate>1000</update_rate>
        <visualize>1</visualize>
        <topic>sensors/imu</topic>
        <imu>
          <topic>sensors/imu</topic>
          <angular_velocity/>
          <linear_acceleration/>
        </imu>
      </sensor>

      <inertial>
        <mass>1000</mass>
        <inertia>
          <ixx>21.66666667</ixx>
          <ixy>0</ixy>
          <iyy>21.66666667</iyy>
          <iyz>0</iyz>
          <izz>41.666667</izz>
        </inertia>
      </inertial>
    </link>

    <link name="wing_sail_link">
      <pose>0 0 0.05 0 0 0</pose>
      <visual name="wing_sail_visual">
        <geometry>
          <mesh>
            <uri>model://land_yacht/meshes/wing_sail.stl</uri>
          </mesh>
        </geometry>
        <material>
          <ambient>1.0 0.4 0.0 0.9</ambient>
          <diffuse>1.0 0.4 0.0 0.9</diffuse>
          <specular>0.1 0.1 0.1 1.0</specular>
        </material>
      </visual>

      <visual name="cp_wing_sail_visual">
        <pose>-0.05 0 0.65 0 0 0</pose>
        <geometry>
          <sphere>
            <radius>0.01</radius>
          </sphere>
        </geometry>
        <material>
          <ambient>1.0 0.0 0.0 1.0</ambient>
          <diffuse>1.0 0.0 0.0 1.0</diffuse>
          <specular>0.1 0.1 0.1 1.0</specular>
        </material>
      </visual>

      <collision name="wing_sail_collision">
        <geometry>
          <mesh>
            <uri>model://land_yacht/meshes/wing_sail.stl</uri>
          </mesh>
        </geometry>
      </collision>

      <inertial>
        <pose>-0.05 0.0 0.65 0 0 0</pose>
        <mass>2.0</mass>
        <inertia>
          <ixx>0.326733333</ixx>
          <ixy>0</ixy>
          <iyy>0.348266667</iyy>
          <iyz>0</iyz>
          <izz>0.021666667</izz>
        </inertia>
      </inertial>
    </link>

    <joint type="revolute" name="wing_sail_joint">
      <pose> 0 0 0 0 0 0</pose>
      <child>wing_sail_link</child>
      <parent>base_link</parent>
      <axis>
        <xyz>0 0 1</xyz>
        <limit>
          <lower>-1.0E16</lower>
          <upper>1.0E16</upper>
        </limit>
        <dynamics>
          <damping>0.01</damping>
        </dynamics>
      </axis>
      <sensor name="force_torque_sensor" type="force_torque">
        <always_on>true</always_on>
        <update_rate>30</update_rate>
        <visualize>true</visualize>
      </sensor>
    </joint>

    <!-- <link name="wing_sail_counterweight_link">
      <pose>0.65 0 0.70 0 0 0</pose>
      <visual name="visual">
        <geometry>
          <sphere>
            <radius>0.025</radius>
          </sphere>
        </geometry>
        <material>
          <ambient>1.0 1.0 0.0 1.0</ambient>
          <diffuse>1.0 1.0 0.0 1.0</diffuse>
          <specular>0.1 0.1 0.1 1.0</specular>
        </material>
      </visual>

      <collision name="collision">
        <geometry>
          <sphere>
            <radius>0.025</radius>
          </sphere>
        </geometry>
      </collision>

      <inertial>
        <mass>0.1</mass>
        <inertia>
          <ixx>0.0025</ixx>
          <ixy>0</ixy>
          <iyy>0.0025</iyy>
          <iyz>0</iyz>
          <izz>0.0025</izz>
        </inertia>
      </inertial>
    </link>

    <joint type="fixed" name="wing_sail_counterweight_joint">
      <child>wing_sail_counterweight_link</child>
      <parent>wing_sail_link</parent>
    </joint> -->

  </model>
</sdf>
